blog: 0.20 release (#1199)
h2oai/wave: website/blog/2022-01-21-release.md

---
slug: release-0.20
title: "New in Wave 0.20"
author: Martin Turoci
author_title: Software Engineer @ H2O.ai
author_url: https://github.com/mturoci
author_image_url: https://avatars.githubusercontent.com/u/64769322?v=4
image: /img/wave-type-yellow-1100x400.png

tags: [release]
---

It's been a long time since the last `0.19` release, but the wait was worth it. The newest Wave includes plenty of extremely useful features that aim to ease the development process and make your Wave apps more appealing. Let's explore the biggest of them.

<!--truncate-->

> ## Bundled Wave server

Reviewer: ```suggestion
## Simpler, faster installation experience
```

> Until the `0.20` version, developers were required to manually download Wave server (waved) from [our Github releases page](https://github.com/h2oai/wave/releases), extract and run before running the app. This step is now eliminated and all you need is to run `pip install h2o_wave`. After successful install, run `wave run your-app` and the Wave server should be started automatically.

Reviewer: ```suggestion
Until the `0.20` version, it was necessary to manually download the Wave server (`waved`) from [our Github releases page](https://github.com/h2oai/wave/releases), extract and run it prior to running your app. This step is no longer necessary. 

Starting with `0.20`, all you need to do is `pip install h2o_wave` and `wave run your-app`. The Wave server is automatically started for you.
```

If you for, some reason, do not want to start `waved` during `wave run`, you can turn it off via `H2O_WAVE_NO_AUTOSTART` env variable.

All the examples and Tour app were previously included in the release. This is not the case with `pip install` though. If you would still like to have all the examples accessible locally, just run `wave fetch` to download examples, demos, and the interactive tour.

```shell
(venv) $ wave fetch
```

You should now have access to 200+ examples locally.

```
Fetching examples and related files. Please wait...
Downloading https://github.com/h2oai/wave/releases/download/v0.19.0/wave-0.19.0-linux-amd64.tar.gz
Extracting...

All additional files downloaded and extracted successfully!
Examples and tour............. /home/elp/git/wave/py/wave/examples
Demos and layout samples...... /home/elp/git/wave/py/wave/demo
Automated test harness........ /home/elp/git/wave/py/wave/test
Wave daemon for deployments... /home/elp/git/wave/py/wave
```

## Base URL support

> By default, Wave apps are all registered on `/` root path. The location your app listens on depends on the route you specify in `@app` or `site['']`. However, if you want to namespace apps or just facing complex deployment, you might want to configure it from a single place rather than on per-app basis.

Reviewer: ```suggestion
By default, Wave apps are all hosted at the root path `/`. This path can be customized by specifying a route via, say, `@app('/foo')` or `site['/foo']`. Starting with v0.20, you can additionally specify a base URL, so that all the apps or sites hosted by a Wave server are served at a path other than `/`. 

Using this feature, specifying a base URL, say `/my/apps`, will make your apps available at `/my/apps/app1`, `/my/apps/app2`, and so on.
```

> Launch `waved` with `-base-url` or launch `waved` with `$H2O_WAVE_BASE_URL` to set a base URL. Setting base URL to `/foo/` will allow serving an app with route `/` at `/foo/`, or an app with route `/bar` at `/foo/bar/`. Accessing `/foo/`'s parent(s) from browser/API will return 404. If a base URL is set, an app must be launched with `$H2O_WAVE_BASE_URL` set to the same base URL value.

Reviewer: ```suggestion
To enable a base URL, launch `waved` with `-base-url` or launch `waved` with `$H2O_WAVE_BASE_URL`. Setting base URL to `/foo/` will allow serving an app with route `/` at `/foo/`, or an app with route `/bar` at `/foo/bar/`. Accessing `/foo/`'s parent(s) from either a browser or API call will return a 404 error. If a base URL is set, an app must be launched with `$H2O_WAVE_BASE_URL` set to the same base URL value.
```

> ## Improved docs

Reviewer: ```suggestion
## Card and component documentation
```

> For the sake of time, our documentation was mostly made up of plain code examples without any extra explanations/descriptions. The [newest version](/docs/components/overview), although not yet fully complete, tries to fix that by organizing the available components according to their usecase and providing all the possible component variations with guidance on when and how they should be used. Additionally, all the documented components include accompanying screenshots so that developers know exactly how is the particular component going to render.

Reviewer: ```suggestion
The documentation now includes a whole new section on [Widgets](/docs/widgets/overview), which includes explanations and examples on how to best use each of the cards and components available in Wave. This is a work in progress, and will be improved over time. We've tried to include screenshots of each example wherever possible, which should make it easier for you to browse, copy and adapt the sample code for your apps.
```

> ## Component additions/improvements

Reviewer: ```suggestion
## New cards, components, and improvements
```

> This release contains a lot of bugfixes and small features that are not worth mentioning from a general overview point of view. Instead, here is a list of the most significant component changes and additions.

Reviewer: ```suggestion
v0.20 ships with more new cards and improvements. Here's an overview of the most important changes.
```

> ### Header

Reviewer: - Is this list complete?
- Would be good to include links to the /widgets/xyz page that describes the widget.

Author: > Is this list complete?

Will check again, but these should be all the new cards/components that were added in 0.20.

> Would be good to include links to the /widgets/xyz page that describes the widget.

Agree, will do.

The header received 3 new attributes:

* `items` - Components that should be displayed on the right side of the header.
* `secondary_items` - Items that should be displayed in the center of the header
* `color` - Header background color. Available colors are 'card' | 'transparent' | 'primary'. Defaults to 'primary'.

![header-improvements](assets/2022-01-21/header.png)

### Sidebar

Sidebar (`ui.nav`) got 4 new attributes:

* `image` - The logo displayed at the top.
* `persona` - The user avatar displayed at the top.
* `secondary_items` - Items that should be displayed at the bottom of the card if items are not empty, otherwise displayed under subtitle.
* `color` - Card background color. Available colors are `'card'` and `'primary'`. Defaults to `'card'`.

![sidebar-improvements](assets/2022-01-21/sidebar.png)

### Footer

A new `items` attribute was added to allow displaying of components (preferably `ui.links`) to the right side of the caption.

![footer-improvements](assets/2022-01-21/footer.png)

### Tall stats card

![tall-stat-card](assets/2022-01-21/tall-stats.png)

> ### Postcard card

Reviewer: ```suggestion
### Post card
```

![postcard-card](assets/2022-01-21/postcard.png)

### Preview card

![preview-card](assets/2022-01-21/preview.png)

### Wide article preview

![wide-article-preview-card](assets/2022-01-21/wide-article.png)

> ## More built-in themes

Reviewer: ```suggestion
## New Themes
```

> If you find choosing the right colors for your Wave app troublesome due to lack of artistic feelings, we have good news! This release introduces 12 new built-in themes to choose from. The inspiration for every theme was taken from some of the most popular code editor/IDE themes.

Reviewer: ```suggestion
v0.20 includes a dozen new built-in themes to choose from. Most of the themes were inspired by popular editor/IDE themes of the same name. If you have ideas for other themes that would be useful to have in Wave, please [let us know](https://github.com/h2oai/wave/discussions)!
```

### Fuchasia

Credit: <https://github.com/thvardhan/Gradianto>.

![fuchasia-theme](assets/2022-01-21/fuchasia.png)

### Monokai

Credit: <https://monokai.pro/>.

![monokai-theme](assets/2022-01-21/monokai.png)

### Nature

Credit: <https://github.com/thvardhan/Gradianto>.

![nature-theme](assets/2022-01-21/nature.png)

### Nord

Credit: <https://www.nordtheme.com/>.

![nord-theme](assets/2022-01-21/nord.png)

### Oceanic

Credit: <https://material-theme.com/>.

![oceanic-theme](assets/2022-01-21/oceanic.png)

### One dark pro

Credit: <https://marketplace.visualstudio.com/items?itemName=zhuangtongfa.Material-theme>.

![one-dark-pro-theme](assets/2022-01-21/one-dark-pro.png)

### Benext

Credit: <https://slackthemes.net/#/benext>.

![benext-theme](assets/2022-01-21/benext.png)

### Ember

Credit: <https://emberjs.com/>.

![ember-theme](assets/2022-01-21/ember.png)

### Kiwi

Credit: <https://rainglow.io/>.

![kiwi-theme](assets/2022-01-21/kiwi.png)

### Lighting

Credit: <https://slackthemes.net/#/lightning>.

![lighting-theme](assets/2022-01-21/lighting.png)

### Solarized

Credit: <https://ethanschoonover.com/solarized/>.

![solarized-theme](assets/2022-01-21/solarized.png)

> Have an idea for a good-looking theme that others might like? Don't hesitate, [start a discussion](https://github.com/h2oai/wave/discussions) with us and it may appear in the next Wave release!

Reviewer: ```suggestion

## Feedback

We look forward to continuing our collaboration with the community and hearing your feedback as we further improve and expand the H2O Wave platform.

We'd like to thank the entire Wave team and the community for all of the contributions to this work!
```
